common/modules/reserve_queue/src/reserve_queue.rs

```diff
@@ -0,0 +1,417 @@
+        _precision: u64,
+        _queues: &mut Vec<u64>,
+    ) -> SCResult<()> {
+        let exists = !self.reserve(token_id, nonce).is_empty();
+
+        if exists {
+            self.reserve(token_id, nonce).update(|x| x.amount += amount);
+        } else {
+            self.reserve(token_id, nonce)
+                .set(&ReserveNode::from(amount.clone(), 0));
+        }
+
+        Ok(())
+    }
+
+    fn push_single_non_fungible(
+        &self,
+        token_id: &TokenIdentifier,
+        nonce: Nonce,
+        amount: &Self::BigUint,
+        precision: u64,
+        queues: &mut Vec<u64>,
+    ) -> SCResult<()> {
+        require!(nonce != 0, "Nonce cannot be zero");
+
+        if self.reserve(token_id, nonce).is_empty() {
+            self.push_single_non_fungible_new(
+                token_id,
+                nonce,
+                amount,
+                precision,
+                queues,
+                Option::None,
+            );
+        } else {
+            self.push_single_non_fungible_existing(token_id, nonce, amount, precision, queues);
+        }
+
+        Ok(())
+    }
+
+    fn should_change_queue(&self, current: u64, new: u64) -> bool {
+        current != new
+    }
+
+    fn get_precision_or_default(&self, token_id: &TokenIdentifier) -> u64 {
+        let precision = self.precision(token_id).get();
+
+        if precision != 0 {
+            precision
+        } else {
+            DEFAULT_PRECISION
+        }
+    }
+
+    fn get_compressed_pairs(
+        &self,
+        pairs: &[NonceAmountPair<Self::BigUint>],
+    ) -> Vec<NonceAmountPair<Self::BigUint>> {
+        let mut pairs_compressed = Vec::<NonceAmountPair<Self::BigUint>>::new();
+
+        pairs.iter().filter(|&x| x.amount != 0).for_each(|pair| {
+            match pairs_compressed.iter().position(|x| x.nonce == pair.nonce) {
+                Some(index) => pairs_compressed[index].amount += &pair.amount,
+                None => pairs_compressed.push(pair.clone()),
+            }
+        });
+
+        pairs_compressed
+    }
+
+    fn get_queue_id(&self, amount: &Self::BigUint, precision: u64) -> u64 {
+        let mut queue_id = 0;
+
+        let precision_biguint = Self::BigUint::from(precision);
+        let mut amount_clone = amount.clone();
+        while amount_clone != 0 {
+            amount_clone /= &precision_biguint;
+            queue_id += 1;
+        }
+
+        queue_id
+    }
+
+    fn push_single_non_fungible_new(
+        &self,
+        token_id: &TokenIdentifier,
+        nonce: Nonce,
+        amount: &Self::BigUint,
+        precision: u64,
+        queues: &mut Vec<u64>,
+        queue_id_opt: Option<u64>,
+    ) {
+        let queue_id = queue_id_opt.unwrap_or_else(|| self.get_queue_id(amount, precision));
+        let head = self.head(token_id, queue_id).get();
+
+        if head == 0 {
+            self.init_queue(token_id, queue_id, nonce, amount);
+            queues.push(queue_id);
+            queues.sort();
```

Comment:
why do you sort when only one token ?

Reply:
Its a vector of queue ids. when a new queue is instantiated, it needs to be added to the vector, but also the vector has to remain sorted. It should not have more than say 30 elements, so should be pretty fast and cheap